File: scheduler_py/hard_constraints.py
import random
from typing import List, Dict, Any, Tuple

from .models import Cabin, ActivityArea, Period, Assignment, DoubleBookingLikelihood
from .utils import (
    get_area_utilization,
    get_last_cabin_area,
    calculate_travel_time,
    has_cabin_used_area_recently,
)

def check_hard_constraints(
    cabin: Cabin,
    area: ActivityArea,
    period: Period,
    assignments: List[Assignment],
    config: Dict[str, Any],
    periods: List[Period]
) -> Tuple[bool, str]:
    """
    Check if a cabin can be assigned to an area during a period based on all hard constraints.
    Returns a tuple of (isValid, reason).
    """
    if is_cabin_already_assigned(cabin.id, period.day, period.id, assignments):
        return False, "Cabin already assigned during this period"

    if not check_area_capacity(area, period.day, period.id, assignments):
        return False, "Area at maximum capacity"

    if not check_area_conflicts(area, period.day, period.id, assignments):
        return False, "Area conflict detected"

    if not check_travel_time_constraints(cabin, area, period, assignments, config, periods):
        return False, "Excessive travel time between areas"

    if not check_fixed_area_closures(area, period):
        return False, "Area closed during this period"

    if not check_no_repeats_rule(cabin.id, area.id, period.day, assignments, config):
        return False, "Cabin used this area too recently"

    if not check_buffer_periods(area, period.day, period.id, assignments):
        return False, "Buffer period required after previous use"

    if not check_cabin_blackout_periods(cabin, period):
        return False, "Cabin blacked out during this period"

    if not check_cabin_blackout_areas(cabin, area):
        return False, "Cabin blacked out from this area"

    return True, "All hard constraints satisfied"

def is_cabin_already_assigned(cabin_id: str, day: int, period_id: str, assignments: List[Assignment]) -> bool:
    """Check if a cabin is already assigned during a specific period."""
    return any(a.cabin_id == cabin_id and a.day == day and a.period_id == period_id for a in assignments)

def check_area_capacity(area: ActivityArea, day: int, period_id: str, assignments: List[Assignment]) -> bool:
    """Check if area capacity constraints are satisfied."""
    current_utilization = get_area_utilization(area.id, assignments, day, period_id)
    return current_utilization < area.max_capacity

def check_area_conflicts(area: ActivityArea, day: int, period_id: str, assignments: List[Assignment]) -> bool:
    """Check for area conflicts (mutually exclusive areas)."""
    if not area.linked_areas:
        return True
    for linked_area_id in area.linked_areas:
        if get_area_utilization(linked_area_id, assignments, day, period_id) > 0:
            return False
    return True

def check_travel_time_constraints(cabin: Cabin, area: ActivityArea, period: Period, assignments: List[Assignment], config: Dict[str, Any], all_periods: List[Period]) -> bool:
    """Check travel time constraints between consecutive periods."""
    last_area_id = get_last_cabin_area(cabin.id, assignments, period.day, period.id, all_periods)
    if not last_area_id:
        return True

    last_area = next((a for a in config.get("areas", []) if a.id == last_area_id), None)
    if not last_area:
        return True

    travel_time = calculate_travel_time(last_area, area)
    max_allowed_time = config.get("allowedTransitionTime", 30)
    return travel_time <= max_allowed_time

def check_fixed_area_closures(area: ActivityArea, period: Period) -> bool:
    """Check if area is closed during the period."""
    return area.id not in period.blackout_areas

def check_no_repeats_rule(cabin_id: str, area_id: str, day: int, assignments: List[Assignment], config: Dict[str, Any]) -> bool:
    """Check no repeats rule."""
    no_repeats_days = config.get("noRepeatsDays", 3)
    return not has_cabin_used_area_recently(cabin_id, area_id, assignments, day, no_repeats_days)

def check_buffer_periods(area: ActivityArea, day: int, period_id: str, assignments: List[Assignment]) -> bool:
    """Check buffer periods after area use."""
    if area.buffer_periods == 0:
        return True
    return not any(a.area_id == area.id and a.day == day and a.period_id != period_id for a in assignments)

def check_cabin_blackout_periods(cabin: Cabin, period: Period) -> bool:
    """Check if cabin is blacked out during the period."""
    return period.id not in cabin.restrictions.blackout_periods

def check_cabin_blackout_areas(cabin: Cabin, area: ActivityArea) -> bool:
    """Check if cabin is blacked out from the area."""
    return area.id not in cabin.restrictions.blackout_areas

def is_double_booking_allowed(area: ActivityArea) -> bool:
    """Check if double booking is allowed for an area."""
    if area.double_booking.likelihood == DoubleBookingLikelihood.NEVER:
        return False
    if area.double_booking.likelihood == DoubleBookingLikelihood.ALWAYS:
        return True
    if area.double_booking.likelihood == DoubleBookingLikelihood.SOMETIMES:
        return random.random() < 0.3  # 30% chance
    return False

def is_area_alternating_days(area: ActivityArea) -> bool:
    """Check if area alternates days and is exempt from other rules."""
    return area.alternates_days

File: src/algorithm/constraints/hardConstraints.js
/**
 * Hard constraint functions for camp scheduling
 * These functions implement the "Programmatic Rules" and hard aspects of "Conditionals"
 */

import { 
  getAreaUtilization, 
  getLastCabinArea, 
  calculateTravelTime,
  hasCabinUsedAreaRecently
} from './utilities.js';

/**
 * Check if a cabin can be assigned to an area during a period
 * @param {Object} cabin - Cabin object
 * @param {Object} area - Activity area object
 * @param {Object} period - Period object
 * @param {Array} assignments - Current assignments
 * @param {Object} config - Configuration object
 * @returns {Object} Result with isValid boolean and reason string
 */
export function checkHardConstraints(cabin, area, period, assignments, config) {
  // Check if cabin is already assigned during this period
  if (isCabinAlreadyAssigned(cabin.id, period.day, period.id, assignments)) {
    return { isValid: false, reason: 'Cabin already assigned during this period' };
  }

  // Check area capacity
  if (!checkAreaCapacity(area, period.day, period.id, assignments)) {
    return { isValid: false, reason: 'Area at maximum capacity' };
  }

  // Check area conflicts
  if (!checkAreaConflicts(area, period.day, period.id, assignments)) {
    return { isValid: false, reason: 'Area conflict detected' };
  }

  // Check travel time constraints
  if (!checkTravelTimeConstraints(cabin, area, period.day, period.id, assignments, config)) {
    return { isValid: false, reason: 'Excessive travel time between areas' };
  }

  // Check fixed area closures
  if (!checkFixedAreaClosures(area, period)) {
    return { isValid: false, reason: 'Area closed during this period' };
  }

  // Check no repeats rule
  if (!checkNoRepeatsRule(cabin.id, area.id, period.day, assignments, config)) {
    return { isValid: false, reason: 'Cabin used this area too recently' };
  }

  // Check buffer periods
  if (!checkBufferPeriods(area, period.day, period.id, assignments)) {
    return { isValid: false, reason: 'Buffer period required after previous use' };
  }

  // Check cabin blackout periods
  if (!checkCabinBlackoutPeriods(cabin, period)) {
    return { isValid: false, reason: 'Cabin blacked out during this period' };
  }

  // Check cabin blackout areas
  if (!checkCabinBlackoutAreas(cabin, area)) {
    return { isValid: false, reason: 'Cabin blacked out from this area' };
  }

  return { isValid: true, reason: 'All hard constraints satisfied' };
}

/**
 * Check if a cabin is already assigned during a specific period
 * @param {string} cabinId - Cabin ID
 * @param {number} day - Day number
 * @param {string} periodId - Period ID
 * @param {Array} assignments - Current assignments
 * @returns {boolean} True if cabin is already assigned
 */
function isCabinAlreadyAssigned(cabinId, day, periodId, assignments) {
  return assignments.some(a => 
    a.cabinId === cabinId && 
    a.day === day && 
    a.periodId === periodId
  );
}

/**
 * Check if area capacity constraints are satisfied
 * @param {Object} area - Activity area
 * @param {number} day - Day number
 * @param {string} periodId - Period ID
 * @param {Array} assignments - Current assignments
 * @returns {boolean} True if capacity constraints are satisfied
 */
function checkAreaCapacity(area, day, periodId, assignments) {
  const currentUtilization = getAreaUtilization(area.id, assignments, day, periodId);
  
  // Check max capacity
  if (currentUtilization >= area.maxCapacity) {
    return false;
  }
  
  return true;
}

/**
 * Check for area conflicts (mutually exclusive areas)
 * @param {Object} area - Activity area
 * @param {number} day - Day number
 * @param {string} periodId - Period ID
 * @param {Array} assignments - Current assignments
 * @returns {boolean} True if no conflicts
 */
function checkAreaConflicts(area, day, periodId, assignments) {
  if (!area.linkedAreas || area.linkedAreas.length === 0) {
    return true;
  }

  // Check if any linked areas are currently in use
  for (const linkedAreaId of area.linkedAreas) {
    const linkedAreaUtilization = getAreaUtilization(linkedAreaId, assignments, day, periodId);
    if (linkedAreaUtilization > 0) {
      return false;
    }
  }

  return true;
}

/**
 * Check travel time constraints between consecutive periods
 * @param {Object} cabin - Cabin object
 * @param {Object} area - Target activity area
 * @param {number} day - Day number
 * @param {string} periodId - Period ID
 * @param {Array} assignments - Current assignments
 * @param {Object} config - Configuration object
 * @returns {boolean} True if travel time is acceptable
 */
function checkTravelTimeConstraints(cabin, area, day, periodId, assignments, config) {
  const lastAreaId = getLastCabinArea(cabin.id, assignments, day, periodId);
  if (!lastAreaId) {
    return true; // No previous assignment to compare
  }

  // Find the last area object
  const lastArea = config.areas.find(a => a.id === lastAreaId);
  if (!lastArea) {
    return true;
  }

  const travelTime = calculateTravelTime(lastArea, area);
  const maxAllowedTime = config.allowedTransitionTime || 30; // Default 30 minutes

  return travelTime <= maxAllowedTime;
}

/**
 * Check if area is closed during the period
 * @param {Object} area - Activity area
 * @param {Object} period - Period object
 * @returns {boolean} True if area is open
 */
function checkFixedAreaClosures(area, period) {
  // Check if area is in period's blackout areas
  if (period.blackoutAreas && period.blackoutAreas.includes(area.id)) {
    return false;
  }

  return true;
}

/**
 * Check no repeats rule
 * @param {string} cabinId - Cabin ID
 * @param {string} areaId - Area ID
 * @param {number} day - Day number
 * @param {Array} assignments - Current assignments
 * @param {Object} config - Configuration object
 * @returns {boolean} True if no repeats rule is satisfied
 */
function checkNoRepeatsRule(cabinId, areaId, day, assignments, config) {
  const noRepeatsDays = config.noRepeatsDays || 3; // Default 3 days
  
  return !hasCabinUsedAreaRecently(cabinId, areaId, assignments, noRepeatsDays);
}

/**
 * Check buffer periods after area use
 * @param {Object} area - Activity area
 * @param {number} day - Day number
 * @param {string} periodId - Period ID
 * @param {Array} assignments - Current assignments
 * @returns {boolean} True if buffer period requirement is satisfied
 */
function checkBufferPeriods(area, day, periodId, assignments) {
  if (!area.bufferPeriods || area.bufferPeriods === 0) {
    return true;
  }

  // Check if area was used recently and needs buffer
  const recentAssignments = assignments.filter(a => 
    a.areaId === area.id && 
    a.day === day && 
    a.periodId !== periodId
  );

  // Simple buffer check - could be enhanced with actual period timing
  return recentAssignments.length === 0;
}

/**
 * Check if cabin is blacked out during the period
 * @param {Object} cabin - Cabin object
 * @param {Object} period - Period object
 * @returns {boolean} True if cabin can be scheduled during this period
 */
function checkCabinBlackoutPeriods(cabin, period) {
  if (!cabin.restrictions || !cabin.restrictions.blackoutPeriods) {
    return true;
  }

  return !cabin.restrictions.blackoutPeriods.includes(period.id);
}

/**
 * Check if cabin is blacked out from the area
 * @param {Object} cabin - Cabin object
 * @param {Object} area - Activity area
 * @returns {boolean} True if cabin can use this area
 */
function checkCabinBlackoutAreas(cabin, area) {
  if (!cabin.restrictions || !cabin.restrictions.blackoutAreas) {
    return true;
  }

  return !cabin.restrictions.blackoutAreas.includes(area.id);
}

/**
 * Check if double booking is allowed for an area
 * @param {Object} area - Activity area
 * @param {string} cabinId - Cabin ID
 * @param {Array} assignments - Current assignments
 * @param {number} day - Day number
 * @param {string} periodId - Period ID
 * @returns {boolean} True if double booking is allowed
 */
export function isDoubleBookingAllowed(area, cabinId, assignments, day, periodId) {
  if (!area.doubleBooking) {
    return false;
  }

  const { likelihood, scope } = area.doubleBooking;
  
  if (likelihood === 'never') {
    return false;
  }

  if (likelihood === 'always') {
    return true;
  }

  if (likelihood === 'sometimes') {
    // Simple probability check - could be enhanced with more sophisticated logic
    return Math.random() < 0.3; // 30% chance
  }

  return false;
}

/**
 * Check if area alternates days and is exempt from other rules
 * @param {Object} area - Activity area
 * @returns {boolean} True if area alternates days
 */
export function isAreaAlternatingDays(area) {
  return area.alternatesDays === true;
}
